---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```


# argyle
An `R` package for import, QC and (some) analysis of genotyping and hybridization-intensity data from Illumina Infinium arrays.

## Dependencies
Effort has been made to keep to a minimum the number of package dependencies, subject to the constraint that I don't want to re-implement from scratch what others have done better.

* `data.table`: really fast and efficient handling of big (multi-GB scale) table-style data with low overhead
* `preprocessCore` (Biodoncuctor): robust quantile normalization routine written in `C`
* `plyr`: generalizations of base-`R`'s `apply()` family
* `reshape2`: easy "flattening" of matrices to dataframes
* `digest`: for computing MD5 checksums to check data integrity

The following are required for some functions but one could get by without them:

* `ggplot2` (and friends): required for the plotting functions
* `corpcor`: required for "fast"-mode PCA

## Installation
Installation of the package directly from Github requires `devtools`.
```{r, eval=FALSE}
library(devtools)

## allow R to look for pacakges in both CRAN and Bioconductor
setRepositories(1:2)

## install from Github source
install_github("andrewparkermorgan/argyle")
```

## Usage
```{r, eval=FALSE}
library(argyle)

data(snps)
geno <- read.beadstudio("sample", snps, in.path = "./")

## see marker map and sample metadata
markers(geno)
samples(geno)

## subset operations: equivalent pairs
subset(geno, chr == "chrX")
geno[ markers(geno)$chr == "chrX", ]

subset(geno, sex == 2, by = "samples")
geno[ ,samples(geno)$sex == 2 ]

## run QC checks and flag samples above thresholds
geno <- run.qc.checks(geno, max.H = 3000, max.N = 5000)
# how many samples fail QC?
summarize.filters(geno)
# remove samples failing QC
filt <- apply.filters(geno)

## compute BAF/LRR for copy-number analysis
data(clusters)
geno.norm <- tQN(geno, clusters = clusters)

## grab intensity data for specific marker(s)
ii <- get.intensity(geno, c("JAX00240610","JAX00240636"))

## grab BAF/LRR
blr <- get.baf(geno.norm, c("JAX00240610","JAX00240636"))
```

## Performance
For a realistic test of `argyle`'s performance, I use an Illumina BeadStudio dataset from the MegaMUGA array for mouse, containing 77808 markers x 96 samples (available from [do.jax.org](http://churchill.jax.org/research/cc/do_data/megamuga/raw/MegaMUGA_22Oct2012/)).  The relevant files, compressed with ZIP, have total size 149 MB.

Testing on my Mac Pro desktop (OS X 10.6.8.10K549, 2x2.26 GHz quad-core Xeon, 24 GB DDR3 RAM) under  `R` 3.1.3:
```{r, echo=FALSE, results="hide", error=FALSE, warning=FALSE, eval=TRUE}
library(devtools)
load_all()
load("~/db/arrays/megamuga/snps.megamuga.Rdata")
```
```{r, eval=TRUE}
system.time( geno <- read.beadstudio("", snps, "./data/MM_sample") )

summary(geno)

print(object.size(geno), units = "Mb")

sessionInfo()
```

The final object is `r format(object.size(geno), units = "Mb")`, and is completely "self-contained" in that sample and marker metadata are stored alongside the genotypes and hybridization intensities.  It is worth noting that if the input files were decompressed they would occupy about 500 MB on disk.

## Interface to [`R/DOQTL`](http://cgd.jax.org/apps/doqtl/DOQTL.shtml)
Genotypes processed with `argyle` can be packaged into a set of `R` objects (bundled in an `*.Rdata` file) suitable for use as input to Dan Gatti's `DOQTL` software.  `DOQTL` performs haplotype reconstruction and genetic mapping (under both linkage/composite-interval and single-marker association models) in multifounder advanced intercross populations.  Its namesake is the Diversity Outbred (DO) mouse population (see [do.jax.org](http://do.jax.org/)).
```{r, eval=FALSE}
## export for DOQTL
export.doqtl(geno, "./doqtl.objects.Rdata")
```

## Interface to [`PLINK`](https://www.cog-genomics.org/plink2/)
Computation on large SNP array genotyping datasets is not a new problem.  Many common operations -- frequency statistics (sample-wise and marker-wise), differentiation statistics ($F_{st}$ et al), homozygosity checks, association testing, multivariate clustering by PCA and MDS -- are implemented efficiently in the `PLINK` package.  The input formats popularized by `PLINK` are now used by other software in population genetics.

This package provides functions to read and write **binary** `PLINK` filesets.  The binary fileset consists of three files:

* `*.fam`: the ``family file'' describing samples (6 columns): family ID, sample ID, mom ID, dad ID, sex (0=unknown, 1=male, 2=female), phenotype (-9=missing)
* `*.bim`: a file describing the marker map (6 columns, at least): chromosome, marker ID, genetic position (cM), physical position (bp), allele 1, allele 2
* `*.bed`: compact binary representation of genotypes using 2 bits per genotype

Note that order matters: genotypes from the `*.bed` file are mapped to samples and markers using order of appearance in the `*.fam` and `*.bim` files.
```{r, eval=FALSE}
## this command produces files 'sample.bed', 'sample.bim' and 'sample.fam'
write.plink(geno, "./sample")
## ... and this one reads them back in
geno <- read.plink("./sample")
```

Also included are thin wrappers around some `PLINK` utilities.  These of course require a working executable named `plink` in the user's path. [**TODO**]